### DISCUSSION TITLE
[gatsby-source-contentful] and useStaticQuery fails when editor changes content

### DISCUSSION THREAD

**Discussion by eddysims** [gatsbyjs/gatsby#33803] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I have a `Micro Content` content type in Contentful that allows an editor to have a piece of content with a link to an asset.

when I use `useStaticQuery` I get the following error

> Cannot query field "references" on type "ContentfulMicroContentValue".

My query looks like this.

```
const query = graphql`
  query SomeQuery {
    content: contentfulMicroContent(key: { eq: "Some Key" }) {
      value {
        raw
        references {
          contentful_id
          slug
        }
      }
    }
  }
`;
```

This is failing because `references` is an optional field. The editor can choose an asset link or not.

Is there some way that I don't have to change this query based on whether or not there is an asset linked in the content?

### Expected Result

Some fallback that doesn't fail whenever an editor change a content field

### Actual Result

Gatsby fails to query and a deploy is required every time an editor makes a change.

### Environment

```shell
System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.13.0 - ~/.nvm/versions/node/v12.13.0/bin/node
    Yarn: 1.13.0 - /usr/local/bin/yarn
    npm: 6.12.0 - ~/.nvm/versions/node/v12.13.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 95.0.4638.54
    Firefox: 91.0.2
    Safari: 15.0
  npmPackages:
    gatsby: ^3.2.1 => 3.2.1 
    gatsby-plugin-google-gtag: ^3.6.0 => 3.6.0 
    gatsby-plugin-graphql-codegen: ^3.0.0 => 3.0.0 
    gatsby-plugin-image: ^1.2.1 => 1.2.1 
    gatsby-plugin-manifest: ^3.2.0 => 3.2.0 
    gatsby-plugin-postcss: ^4.6.0 => 4.6.0 
    gatsby-plugin-react-helmet: ^4.2.0 => 4.2.0 
    gatsby-plugin-sharp: ^3.2.1 => 3.2.1 
    gatsby-plugin-sitemap: ^3.2.0 => 3.2.0 
    gatsby-plugin-social9-socialshare: ^1.0.5 => 1.0.5 
    gatsby-source-contentful: ^5.2.1 => 5.2.1 
    gatsby-source-filesystem: ^3.2.0 => 3.2.0 
    gatsby-transformer-sharp: ^3.2.0 => 3.2.0
```

**Comment by LekoArts** [comment]
Hi!

Please see: https://www.gatsbyjs.com/plugins/gatsby-source-contentful/#restrictions-and-limitations

If not at least one entry is defined it'll not infer the needed shape at the moment. This is a limitation that is being worked on: https://github.com/gatsbyjs/gatsby/pull/30855

You'll want to use the createSchemaCustomization API in the meantime (e.g. https://github.com/gatsbyjs/gatsby/issues/26112#issuecomment-667883439) to manually create the types or you create "dummy content" by adding a page that you exclude that contains everything (or something like https://stackoverflow.com/questions/68898616/how-to-optionally-query-references-in-gatsby-contentful-rich-text)
